gk/fib.c

```diff
@@ -1696,6 +1723,170 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
 	return del_fib_entry_numerical(&prefix_info, gk_conf);
 }
 
+/*
+ * Stack when function starts:
+ *
+ *   |            |
+ * 3 |   gk_conf  |
+ * 2 |    table   |
+ * 1 | prefix_str |
+ *   |____________|
+ */
+static int
+read_grantor_entries(lua_State *l, struct ipaddr *gt_addrs,
+	struct ipaddr *gw_addrs, char *err_buf, size_t err_buf_len)
+{
+	lua_Integer i, tbl_size;
+
+	tbl_size = lua_objlen(l, 2);
+
+	/* Iterate over a table of tables. */
+	for (i = 1; i <= tbl_size; i++) {
+		const char *gt_ip, *gw_ip;
+		int ret;
+
+		/* Get the table at index i. */
+		lua_rawgeti(l, 2, i);
+
+		/*
+		 * Make sure that the item inside
+		 * the table is a table itself.
+		 */
+		if (!lua_istable(l, -1)) {
+			ret = snprintf(err_buf, err_buf_len,
+				"Grantor entry %ld is not a table\n", i);
+			RTE_VERIFY(ret > 0 && ret < (int)err_buf_len);
+			return -1;
+		}
+
+		lua_getfield(l, 4, "gt_ip");
+		lua_getfield(l, 4, "gw_ip");
+
+		gt_ip = luaL_checkstring(l, 5);
+		gw_ip = luaL_checkstring(l, 6);
```

Comment:
`luaL_checkstring()` can raise a Lua error (i.e. calling `luaL_error()`). The other functions may raise Lua errors as well. I think it is safer to call `read_grantor_entries()` with `lua_cpcall()`, so we guarantee correctness now and in future changes of the code. After this change, we can also replace `snprintf()` calls with `luaL_error()`.

Reply:
Calling `lua_cpcall()` may be so troublesome here that the following approach may be simpler:
1. replace the `for` loop in `read_grantor_entries()` with a recursive function `list_gt_gw_pairs()`;
2. Whenever `list_gt_gw_pairs()` finds a pair, it call itself;
3. At some point, `list_gt_gw_pairs()` cannot find any more pair, then it allocates the needed memory;
4. As `list_gt_gw_pairs()` returns from its multiple calls, it save the pair that each instance found to the newly allocated memory.

Reply:
The second approach works because, after step 3, there cannot be any Lua error. But I don't know which approach would produce the simplest code.

Reply:
Ok, a third approach. This one is **much** simpler than both approaches above: allocate arrays `gt_addrs` and `gw_addrs` using `lua_newuserdata()`. Thus, Lua will release them after the work is done or if something goes wrong.